pkg/operators/wasm/misc.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wasm
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/tetratelabs/wazero"
+	wapi "github.com/tetratelabs/wazero/api"
+)
+
+func (i *wasmOperatorInstance) addMiscFuncs(env wazero.HostModuleBuilder) {
+	exportFunction(env, "setAnnotation", i.setAnnotation,
+		[]wapi.ValueType{
+			wapi.ValueTypeI64, // ObjName
+			wapi.ValueTypeI64, // Name
+			wapi.ValueTypeI64, // Value
+		},
+		[]wapi.ValueType{wapi.ValueTypeI32}, // Error
+	)
+}
+
+// setAnnotation sets the value of an annotation on an object.
+// Params:
+// - stack[0] objName
+// - stack[1] name
+// - stack[2] value
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) setAnnotation(ctx context.Context, m wapi.Module, stack []uint64) {
+	objNamePtr := stack[0]
+	namePtr := stack[1]
+	valuePtr := stack[2]
+
+	objName, err := stringFromStack(m, objNamePtr)
+	if err != nil {
+		i.logger.Warnf("setAnnotation: reading string from stack: %v", err)
+		stack[0] = 1
+		return
+	}
+
+	name, err := stringFromStack(m, namePtr)
+	if err != nil {
+		i.logger.Warnf("setAnnotation: reading string from stack: %v", err)
+		stack[0] = 1
+		return
+	}
+
+	val, err := stringFromStack(m, valuePtr)
+	if err != nil {
+		i.logger.Warnf("setAnnotation: reading string from stack: %v", err)
+		stack[0] = 1
+		return
+	}
+
+	i.config.Set(fmt.Sprintf("%s.annotations.%s", objName, name), val)
```

Comment:
This has to be differentiated between setting an annotation on a datasource and a field:

* datasource:
   datasources.%s.annotations.%s
* fields:
   datasources.%s.fields.%s.annotations.%s

Depending on the number of dots in objName.
* no dot: datasource annotation
* 1+ dots: field annotation

See also #3468

Reply:
The idea is to give full control to the users and allow them to set the annotation on anything:

- datasource: `setObjAnnotation("datasources.myds", "fooannotation", "fooval")`
- fields: `setObjAnnotation("datasources.myds.fields.myfield", "fooannotation", "fooval")`
- programs: `setObjAnnotation("programs.ig_fssl_read_e", "attach_to", config.read)` from #3463

Otherwise we'll need to implement few different specific `set{Program,Annotation,Field}Annotation` helpers.

Reply:
But shouldn't id be called setConfig() with just key and value, then?

Reply:
That can be an option, but the user will have to pass the "annotations" suffix. Do we want to be that generic?

Reply:
I'm not sure - that could easily create a hard-to-debug mess 😄

Reply:
> setObjAnnotation("programs.ig_fssl_read_e", "attach_to", config.read)

Can we attach a single bpf program to several targets? This can be useful for: 
- kprobes, e.g. the pwru use case
- uprobes for custom SSL gadgets

With kprobes, the wasm module can read kallsyms to decide how to annotate the bpf program.

But in the case of uprobes, we don't know in advance what are the container images and target programs that are going to run. It would be good to have a way to parse a container filesystem before deciding where to attach the uprobe. This could be at a different place depending on the container image.

This doesn't have to be supported in this PR but it's good to keep that in mind. I'm wondering if this API with annotations is enough for the advanced use cases.

Reply:
The current gadgets using uprobes are attached to libraries:
- libc (`trace_malloc`)
- libssl, libcrypto (`trace_ssl`)

And the path to the libraries (e.g. `/lib64/libc.so.6`) is determined by parsing `/etc/ld.so.cache` in the container. So adding an annotation like `uprobe/libc:malloc` with a similar syntax as the section name (`SEC("uprobe/libc:malloc")`) is good enough.

But for uprobes on executables (such as Golang programs because they don't use dynamic libraries) might need something different. I'd like to attach the uprobe on `crypto/tls.(*Conn).Write` but ideally the executable path (e.g. `/usr/bin/etcd` in one container, `/usr/bin/opa` in another container) would not need to be specified, neither in the section name nor in the annotation. The wasm module might want to have a way to look for Golang programs in the container filesystem.

Reply:
I still think this API is confusing (especially when compared to #3468 - that and this PR should probably be aligned when we name this "set annotation").

I'm also not sure about introducing annotations to metadata sections other than datasources - do you really think it's necessary and we can't just use the keys/values directly? If we don't need them, then I'd really rather go with `setConfig()` (or similarly named) to give the gadget dev all the possibilities, and add a dedicated `setAnnotation()` for manipulating annotations on the datasources more easily - aligned with the `--annotate` param.

The PR itself looks good to me.

/cc @blanquicet

Reply:
Updated it to be SetConfig instead. The annotations one is still TBD but as you mentioned privately, it can be done by calling `SetConfig("foo.bar.annotations.zas", "yesno")`